### ISSUE TITLE
Please support bitmap mode in integritytab

### ISSUE THREAD

**Issue by ethereal26** [systemd/systemd#27587]
### Is your feature request related to a problem? Please describe

Systemd's integritytab doesn't currently support dm-integrity bitmap mode ([https://gitlab.com/cryptsetup/cryptsetup/-/wikis/DMIntegrity](url)). Bitmap mode can provide much better performance in exchange of some small reliability loss. If the devices underneath dm-integrity are SSDs, it can also reduce the amount of writes issued to them by 50%, therefore increasing their lifespan.

### Describe the solution you'd like

It would be great if systemd supported bitmap more for standalone devices in integritytab.

I think a sensible approach would be to add an additional option for the fourth field (for example, mode=journal or mode=bitmap). If no such an option is present in integritytab, the default would be journal mode. This way existing configurations would not be broken.

Changing the open mode from journal to bitmap (or vice versa) is supported, according to Milan Broz (see [https://gitlab.com/cryptsetup/cryptsetup/-/issues/717](url)), so adding this new option would pose no data loss problem for unwary users that change the journal or bitmap setting without being aware of its implications.

### Describe alternatives you've considered

Right now, the only option if you use systemd's integritytab is to set the dm-integrity mapping in journal mode.

### The systemd version you checked that didn't have the feature you are asking for

v253

**Comment by aklomp** [comment]
@ethereal26 A possible workaround could be to use `dm_crypt` with added integrity. `cryptsetup luksFormat` accepts the `--integrity-no-journal` option. I'm using something like this:

```sh
cryptsetup luksFormat \
  --type luks2 \
  --label cryptbackup0 \
  --sector-size 512 \
  --pbkdf argon2i \
  --integrity hmac-sha256 \
  --integrity-no-journal \
  /dev/sdxxxx \
  /etc/luks-keys/cryptbackup
```

According to the [man page](https://man.archlinux.org/man/core/cryptsetup/cryptsetup-luksOpen.8.en) for `cryptsetup luksOpen`, if you manually open the device with the `--persistent --integrity-no-journal` flag, the integrity-no-journal option will be persistently stored in the LUKS2 metadata. So after opening the volume manually once, you should be able to auto-open it with that option from `/etc/crypttab`.

**Comment by ethereal26** [comment]
> @ethereal26 A possible workaround could be to use `dm_crypt` with added integrity. `cryptsetup luksFormat` accepts the `--integrity-no-journal` option. I'm using something like this:
> 
> ```shell
> cryptsetup luksFormat \
>   --type luks2 \
>   --label cryptbackup0 \
>   --sector-size 512 \
>   --pbkdf argon2i \
>   --integrity hmac-sha256 \
>   --integrity-no-journal \
>   /dev/sdxxxx \
>   /etc/luks-keys/cryptbackup
> ```
> 
> According to the [man page](https://man.archlinux.org/man/core/cryptsetup/cryptsetup-luksOpen.8.en) for `cryptsetup luksOpen`, if you manually open the device with the `--persistent --integrity-no-journal` flag, the integrity-no-journal flag will be persistently stored in the LUKS2 metadata. So after opening the volume manually once, you should be able to auto-open it with that option from `/etc/crypttab`.
> 
> Edit: I do support this issue because I ran into the same problem while playing with integrity volumes.

Thanks for your kind suggestion.

The typical use case for standalone `integritysetup/dm-integrity` is different than the one for `cryptsetup/dm-crypt`. For standalone `integritysetup`, you only care about detecting data corruption due to hardware errors, and you can use a fast, non-cryptographic hash function. Standalone `integritysetup` is usually combined with MD raid, to properly detect and repair data corruption due to hardware. On top of MD raid, you can also use `cryptsetup/dm-crypt`, and do the data encryption/decryption only once, instead of multiple times (once per disk).

As you point out, you can use `cryptsetup` with additional integrity protection. But AFAIK, by doing so you are obliged to use encryption and a cryptographic hash function (see [https://man7.org/linux/man-pages/man8/cryptsetup.8.html](url), section "Authenticated disk encryption"). So, there is a performance drop and/or an increase in CPU usage.

AFAIK, `cryptsetup` doesn't support bitmap mode either. The `--integrity-no-journal` option you point out is not the same as the bitmap mode. It would activate the `dm-integrity` mapping in direct mode, with no synchronization between the data and the integrity protection. This mode is unsafe. If there is a system crash or a power loss, most probably there will be read errors due to the integrity not matching the data.

BTW, thanks for supporting this issue. Let's see if there is further interest about it. For me it's important because I have some setups with MD RAID on top of `dm-integrity`, and I would like to have the full performance of the bitmap mode. (I also have the option of using ZFS, with also provides integrity protection with no double writing, but for my use case `MD-RAID+dm-integrity` is a better fit).

**Comment by aklomp** [comment]
Thanks for your quick response. I think we both arrived at this issue because we have the same use case in mind: setting up two or more dm-integrity volumes in order to create an MD RAID array on top of them, to which we can potentially add a dm-crypt layer. I agree with everything you wrote above when it comes to comparing plain dm-integrity with the dm-integrity + dm-crypt solution I outlined. Plain dm-integrity > md-raid > dm-crypt is more flexible (but potentially slower?) than (dm-integrity + dm-crypt) > md-raid, and I'm with you when it comes to preferring that setup. Except for the fact that it's slow when we cannot select bitmap mode in `integritytab`, which is why we're both here.

Thanks for pointing out that `cryptsetup` does not support bitmap mode, because it hadn't occurred to me. I reasoned that disabling the journal would automatically select bitmap mode, but I think you might be right and neither might be enabled. I'm not sure how unsafe it is when coupled with a journaling filesystem and a RAID1 array for a backup volume though: in the best case you lose data if the system goes down, but I think it should still be robust against corruption. I'm fine with incomplete backups if the system goes down during a backup, as long as any backups that were made are guaranteed to be integer.

**Comment by aklomp** [comment]
I think there has been a misunderstanding. Where I wrote `(dm-integrity + dm-crypt) > md-raid`, I meant that `(dm-integrity + dm-crypt)` comes before (sits below) the `md-raid` layer. I guess due to the ambiguity of the `>` symbol, the opposite meaning was conveyed; sorry about that. I fully agree with your assessment of how to structure the topology, with `dm-integrity` as the bottom layer in the stack. I think we're after the same goal: a self-healing volume where some flavor of `dm-integrity` provides an integrity-checking resilience layer below an `md-raid` volume.

Thanks for expanding on the safety comment. In my case I think I can live without the journal because I'm using `rsnapshot` to create snapshots on the RAID volume, which is formatted with journaled `ext4`. Starting a snapshot is a manual action, and if the snapshot fails midway due to power outages or other factors, I'll delete it manually and rerun the script. This is not my only backup, rather it's a fun way to use two old 500GB disks that I have sitting around to do something useful. So I'm honestly not too bothered about write-time data loss, as long as it's detectable. I'm way more bothered about bit rot and encryption at rest.

All that aside, about the issue at hand. I looked at [the systemd code](https://github.com/systemd/systemd/blob/main/src/integritysetup/integrity-util.c#L40) that translates `integritytab` into actions, and I think it should be easy to add support for a bitmap-only keyword. If we introduce, say, `no-journal-bitmap` for that, we can set [the proper flag for bitmap-without-journal](https://github.com/mbroz/cryptsetup/blob/main/src/integritysetup.c#L344) like this:

```c
} else if (streq(word, "no-journal-bitmap")) {
        if (ret_activate_flags)
                *ret_activate_flags |= CRYPT_ACTIVATE_NO_JOURNAL_BITMAP;
```
That's the extent of the change needed, although we might want to check that this option isn't mutually exclusive with other options. I like your suggestion for `mode=(journal|bitmap|none)`, and it would also be easy to implement, but not quite in three lines.

**Comment by ethereal26** [comment]
I suggested `mode=(journal|bitmap|direct)` as the new option to add because it's easy to remember and very intuitive for sysadmins. It would be quite easy for them to know the semantics of that option. The `no-journal-bitmap` you suggest could also be a good choice. I think it's perhaps a bit less sysadmin friendly as it has both the "journal" and "bitmap" words, and that can be a bit confusing. It's more developer friendly though, as it's more consistent with integritysetup source code internals.

I tend to prefer the first one a bit more, but it's not a huge deal for me. I think that if properly documented in the man page, any of them can be a good option.

One more thing... I see there is another issue opened ([https://github.com/systemd/systemd/pull/27824](url)), this one for setting the mapping in direct mode. It already has a commit that takes the second approach. As that issue is related with this one, perhaps we could add some comment there, to get other's feedback.

**Comment by aklomp** [comment]
I also prefer your `mode=` suggestion, but at this point `no-journal` has already made it to the mainline (albeit just one week ago), so it has set a precedence. Perhaps we should request that the other issue be reopened and the commit be reverted, so that our more comprehensive proposal can be implemented. My feeling is that the other issue was implemented in a rushed and ad-hoc way.

I'm unfamiliar with how systemd is governed though, so I don't know if rollbacks are a realistic option.

**Comment by ethereal26** [comment]
I agree with all your points.

I don't know either how systemd is managed, so the only way to know would be perhaps to drop some comment on the other issue or ask for reopening it as you suggest.

**Comment by aklomp** [comment]
I forked systemd and pushed a [branch](https://github.com/aklomp/systemd/commits/issue-27587) to implement `mode=(journal|bitmap|direct)` and deprecate `no-journal`. I didn't actually test the code, but it's trivial enough. I guess we could open a PR for this and see what everyone thinks.

**Comment by ethereal26** [comment]
Great!!

**Comment by poettering** [comment]
by all means send a PR we can review

**Comment by aklomp** [comment]
Pushed a PR (#27973)

**Comment by aklomp** [comment]
Thanks for the blog link. It got me thinking (and reading) about bitmap mode. The way I assumed it would work is that the bitmap acts as a kind of atomic commit log. I expected that a bit set inside the bitmap would indicate that that sector and its tag have been fully committed. If the machine crashes during a write, any sectors with unset bits would be discarded. Data integrity above everything else, right?

But as you point out, that's not what actually happens. What happens is that a bit is set for each written sector to indicate that its tag needs a recalculation at some future point. That leaves a huge gap for corruption to happen. Not only can the machine crash and corrupt the data on the disks, after which the bad data will be baked in as you point out, but also, if I understand this right, the tag is actually calculated from the on-disk contents and not from the data sent by the kernel! (I haven't read the code so I don't know if that's actually the case, but I don't see how else you could do deferred checksum calculations.) That would mean that if you write to a bad sector, you might be recalculating an incorrect tag and also baking in corruption. I really hope I'm misunderstanding something here.

You might be right in concluding that no-journal mode is the most practical. Data and tags are written separately, but at least when the machine crashes, they won't match up and integrity is preserved at the cost of losing some data. I'm happy with that behavior for my use case.

**Comment by ethereal26** [comment]
> Just as a note, looks like I too am going down the same dm-integrity + raid route you're discussing here. I've been doing a lot of investigation into how dm-integrity works, and documenting it at https://github.com/eatnumber1/russ.har.mn/blob/master/machinae/_posts/2023-5-29-i-just-wanted-raid.md. I've got a couple patches out to update the kernel documentation on the subject. That'll eventually turn into a blog post on https://russ.har.mn/machinae/ once I've got everything working

Hi,

You are the guy who contributed that PR for integritytab, aren't you?. I hope you are ok with us providing another PR that has changed your patch a bit. We both thought that was the most comprehensive approach.

Thanks for sharing your investigation on md-raid + dm-integrity. I've made a quick reading of it and it looks very interesting. When I do a proper reading of it, I would like to give you some comments about it. I think it would be better to keep this thread on topic, so maybe we could have contact by e-mail or reddit, if you are ok with that.

**Comment by ethereal26** [comment]
> Thanks for the blog link. It got me thinking (and reading) about bitmap mode. The way I assumed it would work is that the bitmap acts as a kind of atomic commit log. I expected that a bit set inside the bitmap would indicate that that sector and its tag have been fully committed. If the machine crashes during a write, any sectors with unset bits would be discarded. Data integrity above everything else, right?
> 
> But as you point out, that's not what actually happens. What happens is that a bit is set for each written sector to indicate that its tag needs a recalculation at some future point. That leaves a huge gap for corruption to happen. Not only can the machine crash and corrupt the data on the disks, after which the bad data will be baked in as you point out, ~but also, if I understand this right, the tag is actually calculated from the on-disk contents and not from the data sent by the kernel! (I haven't read the code so I don't know if that's actually the case, but I don't see how else you could do deferred checksum calculations.) That would mean that if you write to a bad sector, you might be recalculating an incorrect tag and also baking in corruption.~ I really hope I'm misunderstanding something here. Edit: that would be a really bad design and is probably not how it works at all. It probably just does the same thing as direct mode, except it also populates the bitmap which is used for post-crash "recovery". Nevermind. Still a dangerous mode though.
> 
> You might be right in concluding that no-journal mode is the most practical. Data and tags are written separately, but at least when the machine crashes, they won't match up and integrity is preserved at the cost of losing some data. I'm happy with that behavior for my use case.

Just a brief comment. The bitmap mode works similarly to md-raid with an intent bitmap. It maintains a bitmap of active or hot drive areas, where the data and the integrity are expected to differ. If the system crashes, the data from the hot areas is read and its integrity tags regenerated.

The bitmap mode is pretty safe. The bitmap is continuously being updated, so that the hot areas are only a small percentage of the total drive size. As any modern drive already has some ECC data protection scheme, the odds of having undetected corruption with the bitmap mode is pretty low. The journal mode is safer here, but it comes with a huge performance penalty. Depending on your particular case, and what matters most to you, you have to favor speed or safety.

For you particular case (incremental backups with rsnapshot), the direct mode may work fine (I'm not 100% sure about it though). But for a more generic workload, the bitmap mode is the recommended and safest option. Or the journal mode, albeit with the huge performance penalty already mentioned.

In order to keep this thread on topic, I propose you the same. Maybe we can have contact by e-mail, reddit or any other platform of choice.

BTW, big thanks for writing the PR. Now we'll have that functionality on the next systemd release. ;)
